### PR TITLE
ibmi: return EISDIR instead of EOPNOTSUPP on read from directory fd 

### LINKED ISSUE #2147: ibmi: uv__fs_read of directory returns EOPNOTSUPP instead of EISDIR, breaking global node installs
* **Version**: 1.23.2
* **Platform**: OS400 wernstrom 2 7 0010000ACBAA Os

https://github.com/libuv/libuv/pull/2025 removed the directory check for AIX, which also affects the IBM i platform. On IBM i, libuv runs in PASE, which is an AIX-like environment, but unlike AIX which supports reading from directories fds, PASE does not. Also, unlike other platforms that don't support reading from directory fds, PASE returns `EOPNOTSUPP` instead of `EISDIR`. The net result, is that Node.js built against libuv 1.23.2 or later is no longer able to install global npm modules. More details provided here: https://github.com/nodejs/node/issues/25433

### PR DESCRIPTION
**Pull request by kadler**
On IBM i PASE, EOPNOTSUPP is returned when reading a directory instead
of EISDIR, like (seemingly) every other platform which doesn't support
reading directories. To ensure compatibility with software expecting
EISDIR, we map the EOPNOTSUPP to EISDIR when the fd passed in was a
directory.

This is a partial revert of 25a3894, but scoped to PASE and the fstat
call is moved to the end so it's out of the fast path.

Fixes #2147

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -334,6 +334,10 @@ test_run_tests_CFLAGS += -D_UNIX03_THREADS \
                          -qFLOAT=IEEE
 endif
 
+if OS400
+test_run_tests_CFLAGS += -D_ALL_SOURCE -D_XOPEN_SOURCE=500 -D_LINUX_SOURCE_COMPAT
```
Comment (bnoordhuis) [inline Makefile.am]: This is slightly over the 80 char limit but I'll fix that up when I land it.

**Comment by cjihrig** [comment]
Landed in https://github.com/libuv/libuv/commit/91ca6788bbed45e758ee2744a61f8cf455932763, https://github.com/libuv/libuv/commit/8d62d933ef31fd47ae269395364a280a12eb9636, and https://github.com/libuv/libuv/commit/ca02c36c6ddcf1f74215a63b3fbff04c1a79a1c0 with the long line fixed. Thanks!

**Comment by cjihrig** [comment]
698d46e2d21e0a1e3bd7c65796df8cd40300caba (ibmi: add support for IBM i to autotools build) broke SmartOS in the CI:

```
configure: creating ./config.status
./config.status[12]: export: SHELL##: is not an identifier
```

**Comment by kadler** [comment]
@cjihrig I can't see anything in 698d46e that would have caused that failure to occur. Is there any way to see the full CI log?

**Comment by cjihrig** [comment]
@kadler do you have access to https://ci.nodejs.org/job/libuv-test-commit-smartos/1174/

**Comment by kadler** [comment]
I got access to the CI. Looking at the logs doesn't reveal much. The SmartOS 14 builds worked fine, just the 15 and 16 builds failed.

I don't have a SmartOS system set up to test with, but on my Linux system, line 11-15 of config.status are this:

```shell
SHELL=${CONFIG_SHELL-/bin/sh}
export SHELL
## -------------------- ##
## M4sh Initialization. ##
## -------------------- ##
```
I wonder if somehow the newline after `export SHELL` didn't get written, so it says `export SHELL##` instead. When I run `export SHELL##` from bash I get a similar error.

I have no clue how my changes would have caused that, though.

**Comment by ThePrez** [comment]
Who owns the next action on this?

**Comment by bnoordhuis** [comment]
@ThePrez You or @kadler.

**Comment by ThePrez** [comment]
I'm not sure where to begin trying to correct the SmartOS problem. Can anyone postulate on how our changes may have caused the SmartOS failure? (@libuv/smartos perhaps?)

**Comment by cjihrig** [comment]
I reinstated the commits that were reverted and ran the SmartOS CI again: https://ci.nodejs.org/view/libuv/job/libuv-test-commit-smartos/1239/

Same results. @refack can you give me access to one of the SmartOS instances in the CI?

**Comment by cjihrig** [comment]
Small update. It's not a true solution, but it seems that running `CONFIG_SHELL=/bin/bash ./configure` works around the issue. See https://smartos.org/bugview/OS-7660 for more information.

**Comment by kadler** [comment]
Wow, that's quite obscure!

@cjihrig So how do we proceed from here? Seems that it's been determined that this change is not the root cause of the problem, but if it reliably triggers the problem and that will completely break the CI are we stuck waiting on a ksh93 fix or has the CI been updated with the workaround?

**Comment by cjihrig** [comment]
I think if the couple CI jobs that build libuv on SmartOS are updated, and everything comes back green, we could just add a note in the Build Instructions section of the README and re-land these commits.

@refack is it feasible to update the CI as mentioned in https://github.com/libuv/libuv/pull/2148#issuecomment-472857191?

**Pull request closed without merging**